src/cargo/core/package_id.rs

```diff
@@ -64,51 +87,46 @@ impl<'de> de::Deserialize<'de> for PackageId {
         };
         let source_id = SourceId::from_url(url).map_err(de::Error::custom)?;
 
-        Ok(PackageId {
-            inner: Arc::new(PackageIdInner {
+        Ok(PackageId::wrap(
+            PackageIdInner {
                 name: InternedString::new(name),
                 version,
                 source_id,
-            }),
-        })
-    }
-}
-
-impl Hash for PackageId {
-    fn hash<S: hash::Hasher>(&self, state: &mut S) {
-        self.inner.name.hash(state);
-        self.inner.version.hash(state);
-        self.inner.source_id.hash(state);
+            }
+        ))
     }
 }
 
 impl PartialEq for PackageId {
     fn eq(&self, other: &PackageId) -> bool {
+        if ptr::eq(self.inner, other.inner) {
+            return true;
+        }
         (*self.inner).eq(&*other.inner)
```

Comment:
I don't understand this part. If `ptr::eq` then we have `full_eq` so we `return true`, so far so good. Then we use the `eq` defined on the `PackageIdInner`, but that is the `full_eq` defined above. I would think we would need:

```rust
    fn eq(&self, other: &PackageId) -> bool {
        if ptr::eq(self.inner, other.inner) {
            return true;
        }
        self.inner.name == other.inner.name
            && self.inner.version == other.inner.version
            && self.inner.source_id == other.inner.source_id
    }
```
and that we would need to override hash as well. What am I missing?

Reply:
You're right.  We need to redo the un-done custom do.  Something something different types.